File: leetcode/easy/258-addDigits.js
// Given a non-negative integer num, repeatedly add all its digits until the result has only one digit.

// Example:

// Input: 38
// Output: 2 
// Explanation: The process is like: 3 + 8 = 11, 1 + 1 = 2. 
//              Since 2 has only one digit, return it.

// first pass solution
var addDigits = function(num) {
    while (String(num).length > 1) {
        num = eval(String(num).split("").join("+"));
    }
    return Number(num);
};

// second pass solution
var addDigits = function(num) {
    while (String(num).length > 1) {
        num = String(num);
        let sum = 0;
        
        for (let n of num) {
            sum += parseInt(n);
        }
        num = sum;
    }
    return parseInt(num);
};

// third pass recursive solution
var addDigits = function(num) {
    let sum = 0;
    
    for (let n of String(num)) {
        sum += parseInt(n);
    }
    
    if (String(sum).length > 1) {
        return addDigits(sum);
    } 
    return sum;
};

File: leetcode/easy/258-addDigits.cpp
class Solution {
public:
    int addDigits(int num) {
        string s = to_string(num);
        int sum = 0;

        while (s.length() > 1) {
            for (int i = 0; i < s.length(); i++) {
                cout <<int(s[i] - 48);
                sum += int(s[i] - 48);
            }
            s = to_string(sum);
            sum = 0;
        }

       return stoi(s);
    }
};
